### ISSUE TITLE
ActionDispatch::Routing::RouteSet#recognize_path (assert_recognize) leaks parameters

### ISSUE THREAD

**Issue by libc** [rails/rails#28398]
`ActionDispatch::Routing::RouteSet#recognize_path` does not reset parameters between trying different constraints out. It only [adds](https://github.com/rails/rails/blob/11f10840936a323eda1a9eab02521eff9bdccedb/actionpack/lib/action_dispatch/routing/route_set.rb#L863-L864) parameters. Normal matching doesn't exhibit  the same behavior and there's [a test to make sure of it](https://github.com/rails/rails/blob/11f10840936a323eda1a9eab02521eff9bdccedb/actionpack/test/dispatch/routing_test.rb#L4177-L4195)

### Steps to reproduce

Add the following code to `actionpack/test/controller/routing_test.rb`
```ruby
  def test_recognize_path_does_not_leak_params
    rs.draw do
      get ':foo/bar' => 'bar#foo', constraints: lambda { |*| false }
      get 'foo/:bar' => 'foo#bar'
    end

    assert_equal({ :controller => "foo", :action => 'bar', :bar => 'bar' }, rs.recognize_path("/foo/bar"))
  end
```

As you can see the `:foo` parameter was leaked from the first route with non-matching constraint.

### Expected behavior
No failure

### Actual behavior
```
  1) Failure:
LegacyRouteSetTests#test_recognize_path_does_not_leak_params [test/controller/routing_test.rb:673]:
--- expected
+++ actual
@@ -1 +1 @@
-{:controller=>"foo", :action=>"bar", :bar=>"bar"}
+{:controller=>"foo", :action=>"bar", :foo=>"foo", :bar=>"bar"}
```

### System configuration
**Rails version**: 5.0.2, master

**Ruby version**: 2.3.3

**Comment by maclover7** [comment]
cc @pixeltrix

**Comment by pixeltrix** [comment]
Backported the fix to 5-0-stable in 090adf5efde86b6ed3bb61054822f1135f9b18ba

It appears that 4-2-stable isn't affected - probably due to duping of params somewhere.
